### DISCUSSION TITLE
How does HeightFieldShape create width and height?

### DISCUSSION THREAD

**Discussion by anugentc** [DanielChappuis/reactphysics3d#309] [Q&A] [answered]
Hello! I am trying to get the HeightFieldShape to work for my heightmap. I figured out 95% of it, but one little thing is making me struggle. 

When I create the HeightFieldShape, I pass all of the data I load into it. So I know for a fact that the y points are the same ones as my rendered y points. And that both my render and collision shape is centered at 0, 0, 0. But for some reason my width and height (depth) is off, my X and Z.

I am using a 200 x 200 image to calculate the heightmap. So for my rendered heightmap, I get numbers like -99, 16, -100. But for some reason, the  HeightFieldShape is reporting -98.5, 16, -99.5. 

I am using the physics common createHeightFieldShape() method. Everything else is working well.

If you know anything about why this could be happening, please let me know

![error1](https://user-images.githubusercontent.com/22622013/212178174-e98fd929-fca9-4738-ac2a-ccb66214baf4.PNG)
![error2](https://user-images.githubusercontent.com/22622013/212178177-56bd7b4c-fca5-41b1-92c1-0e42a4f5aed9.PNG)

**Comment by anugentc** [comment, marked as answer]
Alright I think I found something.
So when you create the HeightFieldShape, you subtract 1 from it and divide by 2 to get widths and heights. 

`  mWidth(static_cast<decimal>(nbGridColumns - 1)), mLength(static_cast<decimal>(nbGridRows - 1))`

My image is 200 x 200. So 200 gets put in for the width and height. Subtracting 1 makes it 199.

`
        mAABB.setMin(Vector3(-mWidth * decimal(0.5), -halfHeight, -mLength * decimal(0.5)));
        mAABB.setMax(Vector3(mWidth * decimal(0.5), halfHeight, mLength * decimal(0.5)));`

And here I see that the width and height (length) are set by halving it. So 199 / 2 = 99.5. 

This explains where those values come from. 
But now I need to figure out how to correct it.
If I use 201 then the whole thing gets all jumbled up but I'm not too sure why.

**Reply by DanielChappuis** [reply]
Hello, I am not sure to understand what you are trying to correct. 

> Alright I think I found something. So when you create the HeightFieldShape, you subtract 1 from it and divide by 2 to get widths and heights.
> 
> ` mWidth(static_cast<decimal>(nbGridColumns - 1)), mLength(static_cast<decimal>(nbGridRows - 1))`
> 
> My image is 200 x 200. So 200 gets put in for the width and height. Subtracting 1 makes it 199.


Consider that you have a square grid of 5x5 points where points are spaced by 1 meter. Then, the width of the side of the square is 4 meters. Therefore, we have `width = (nbPoints - 1) * distanceBetweenTwoPoints`. So if you have 200 points in the side of your heightfield, the width is indeed 199 (if the points are spaced by 1).

> 
> ` mAABB.setMin(Vector3(-mWidth * decimal(0.5), -halfHeight, -mLength * decimal(0.5))); mAABB.setMax(Vector3(mWidth * decimal(0.5), halfHeight, mLength * decimal(0.5)));`
> 
> And here I see that the width and height (length) are set by halving it. So 199 / 2 = 99.5.


This line compute the AABB of the whole heightfield. It does not divide the width by two but it compute the extent (half width) of the AABB which is in fact the width of the heightfield divided by two.

**Reply by anugentc** [reply]
Alright, that actually helps a lot. I understand how the spacing works now. 

Looking through the code, I think the problem is yet again my render terrain. When I calculate it i use -(width*/2.0f) + j. So at j=0, it will be at -100. But at j= (width-1) it would be at -(200/2.0f) + 199 , which would be 99.

So I think my render terrain is off. I'm gonna mess around with it and see if I can fix it.

**Reply by anugentc** [reply]
YES! That was it!!

So it was because I was doing that. I changed it to  -((width-1)*/2.0f) + j. This gave me the same values of the HeightFieldShape, and both of them lined up perfectly.

![yes](https://user-images.githubusercontent.com/22622013/212549975-e6750a4e-a1a7-4866-9c82-bfe5c75e260f.PNG)

Your explanation helped a lot, otherwise I would've been chasing the wrong thing. Thank you!

**Reply by DanielChappuis** [reply]
No problem. I am glad that you have found the issue.
